singleheader/simdjson.cpp

```diff
@@ -5771,10 +5770,31 @@ struct value128 {
   uint64_t high;
 };
 
+#if defined(_MSC_VER) && !defined(_M_X64) // _umul128 for x86, arm, arm64
+#if defined(_M_ARM)
+static inline uint64_t __emulu(uint32_t x, uint32_t y) {
+  return x * (uint64_t)y;
+}
+#endif
+static inline uint64_t _umul128(uint64_t ab, uint64_t cd, uint64_t *hi) {
+  uint64_t ad = __emulu((uint32_t)(ab >> 32), (uint32_t)cd);
+  uint64_t bd = __emulu((uint32_t)ab, (uint32_t)cd);
+  uint64_t adbc = ad + __emulu((uint32_t)ab, (uint32_t)(cd >> 32));
+  uint64_t adbc_carry = !!(adbc < ad);
+  uint64_t lo = bd + (adbc << 32);
+  *hi = __emulu((uint32_t)(ab >> 32), (uint32_t)(cd >> 32)) + (adbc >> 32) + (adbc_carry << 32) + !!(lo < bd);
+  return lo;
+}
+#if defined(_M_ARM64)
+uint64_t umul128(uint64_t ab, uint64_t cd, uint64_t* hi) { // non-inline extern version for arm64
+  return _umul128(ab, cd, hi);
```

Comment:
Good.

Reply:
Maybe not so good. You are calling the 32-bit emulated version even if you have _umul128? https://github.com/miranda-ng/miranda-ng/blob/9aa5329a4b4077970571f40f79486a3a1bf3e9f8/libs/libmdbx/src/test/utils.h#L53